### ISSUE TITLE
interp: TestRunnerTerminalExec/Pseudo flaky on Mac

### ISSUE THREAD

**Issue by mvdan** [mvdan/sh#513]
```
--- FAIL: TestRunnerTerminalExec (0.00s)
    --- FAIL: TestRunnerTerminalExec/Pseudo (0.14s)
##[error]        terminal_test.go:129: signal: hangup
FAIL
FAIL	mvdan.cc/sh/v3/interp	4.548s
```

@theclapp perhaps you could help me reproduce this again? Do you have any ideas?

Worst case scenario, we can temporarily skip the pty tests on Mac.

**Comment by mvdan** [comment]
Stress command I tried:

    go test -vet=off -c -o test && stress -p=256 ./test -test.run TestRunnerTerminalExec/Pseudo

Still no failures on Linux, even after 50k or so runs.

**Comment by theclapp** [comment]
-p256 => `ERROR: pipe: too many open files`
Ditto -p128
-p64 => 200k runs, ~16.5m, 0 errors.

**Comment by mvdan** [comment]
A very related flake: https://github.com/mvdan/sh/runs/483406577

```
 --- FAIL: TestRunnerRun (0.03s)
    --- FAIL: TestRunnerRun/817 (0.05s)
##[error]        interp_test.go:2667: wrong output in "echo foo bar > >(sed 's/o/e/g')":
            want: "fee bar\n"
            got:  ""
FAIL
```

**Comment by mvdan** [comment]
And again, the original flake: https://github.com/mvdan/sh/runs/507984876

@theclapp if you have the chance to try again, could you add `-race` to the `go test -c` line above? I still couldn't repro on Linux, even after ~30k runs.

**Comment by theclapp** [comment]
```sh
go test -race -vet=off -c -o test &&
stress -p=80 ./test -test.run TestRunnerTerminalExec/Pseudo
```

66k runs, 0 failures.

**Comment by mvdan** [comment]
@theclapp thanks again. This seems to point to some sort of discrepancy with the Mac machines that GitHub have. Maybe it's a different Mac version? Maybe it's under extra system load? Maybe the setup is weird or buggy? I don't know.

**Comment by theclapp** [comment]
fwiw:
```
% uname -a
Darwin larrymbp14.local 18.7.0 Darwin Kernel Version 18.7.0: \
Thu Jan 23 06:52:12 PST 2020; \
root:xnu-4903.278.25~1/RELEASE_X86_64 x86_64 i386 MacBookPro11,3 Darwin
```

Mojave, 10.14.6 (18G3020)
2.8 GHz Intel Core i7

```
% go version
go version go1.13.4 darwin/amd64
```

**Comment by mvdan** [comment]
Failed once again, this time on a PR: https://github.com/mvdan/sh/pull/531/checks?check_run_id=538743036

It's starting to affect our ability to rely on CI, so I'll start skipping the test. We should keep this issue open to investigate the true cause. My working theory is still that there's something different about Github's Mac machines.

**Comment by myitcv** [comment]
We totally skipped the Mac tests because it was flaking so regularly (at least we attributed the flakes to this issue we're trying to describe/understand)

Perhaps we raise an issue on https://github.com/actions/runner as a next step, if only to centralise discussion?

**Comment by leitzler** [comment]
Just to add another data point:
```
$ uname -a
Darwin MacBook-Pro.localdomain 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64
$ go version
go version go1.14.1 darwin/amd64
```

```
$ go test -vet=off -c -o test && stress -p=64 ./test -test.run TestRunnerTerminalExec/Pseudo
...
202421 runs so far, 0 failures
```

macOS Catalina 10.15.4 - MacBook Pro (13" 2016, 3.3GHz i7, 16GB)

**Comment by mvdan** [comment]
Another similar failure:

```
--- FAIL: TestRunnerRun (0.03s)
    --- FAIL: TestRunnerRun/817 (0.00s)
##[error]        interp_test.go:2668: wrong output in "cat <(cat <(echo nested))":
            want: "nested\n"
            got:  ""
FAIL
```

@theclapp if you have five minutes, could you please try `stress` on this one? The Actions mac machine is upset with these process redirection tests in the past few weeks. It's gotten worse, for some reason. The tests continue to be very solid on Linux, which is all that I can test on right now. With `go test -vet=off -c -o test && stress -p=64 ./test -test.run TestRunnerRun$/817$` I got to 100k runs without a single failure, for example.

**Comment by mvdan** [comment]
Or, if any of you are on mac and are gracious enough to run the more general `go test -vet=off -c -o test && stress -p=64 ./test`, maybe that will do a better job at catching flakes. A smaller `-p` might be needed, though.

**Comment by theclapp** [comment]
I'm actually getting test failures in TestRunnerRunConfirm/612 - 617.  Here's 612, the others are similar:

```
    --- FAIL: TestRunnerRunConfirm/612 (0.03s)
        interp_test.go:2936: wrong bash output in "echo 'set -- g f'>b; echo 'set -- d e f; echo $@; source b;' >a; source a; echo $@":
            want: "d e f\ng f\n"
            got:  "d e f\n\ad e f\n"
```

This may have to do with the particular version of bash on my laptop.  I didn't used to get these failures, I think a "brew upgrade" or "brew install" of bash broke things.  I haven't had an opportunity to try to delve further yet.

**Comment by theclapp** [comment]
I tried the more limited test and got 9 failures in 37 seconds.

Most of them look like this:

```
1204 runs so far, 0 failures

/var/folders/32/8nd008752lv7v3vsgzlrw6qr0000gn/T/go-stress-20200424T090339-429252386
--- FAIL: TestRunnerRun (0.22s)
    --- FAIL: TestRunnerRun/817 (0.01s)
        interp_test.go:2668: wrong output in "cat <(cat <(echo nested))":
            want: "nested\n"
            got:  "file exists\n"
FAIL


ERROR: exit status 1
```

Some of them look like this, where the "got" includes "exit status 1".

```
6564 runs so far, 6 failures

/var/folders/32/8nd008752lv7v3vsgzlrw6qr0000gn/T/go-stress-20200424T090339-274766360
--- FAIL: TestRunnerRun (0.25s)
    --- FAIL: TestRunnerRun/817 (0.01s)
        interp_test.go:2668: wrong output in "cat <(cat <(echo nested))":
            want: "nested\n"
            got:  "file exists\nexit status 1"
FAIL


ERROR: exit status 1
```

**Comment by mvdan** [comment]
Ah, those are the tests which run the real `bash`, not our interpreter. That's what the `Confirm` suffix intends to say, anyway. You can use `-run TestRunnerRun$` to make sure you only run the tests with our interpreter.

**Comment by mvdan** [comment]
Thanks so much @theclapp. So at least the `RunnerRun` flakes are reproducible locally. I don't think they are the same bug as the `hangup` weirndess we're seeing on CI.

**Comment by mvdan** [comment]
I've split off the named pipe (process substitution) flakes into https://github.com/mvdan/sh/issues/576, since they seem unrelated.

**Comment by mvdan** [comment]
All those "file exists" errors with `<(` are a really silly mistake on my part, in hindsight. See 3cb88018ef591cc0b278dd91e75644320870d6da.

The bug and fix seem unrelated to the other flakes mentioned in this thread, though, so I'm keeping this open for now.

**Comment by mvdan** [comment]
Also, for some reason, I now seem to be able to predictably reproduce the pseudo-terminal failures on Linux:

```
$ go test -short -race -count=100 -failfast
--- FAIL: TestRunnerTerminalExec (0.00s)
    --- FAIL: TestRunnerTerminalExec/Pseudo (0.02s)
        unix_test.go:130: signal: hangup
FAIL
exit status 1
FAIL	mvdan.cc/sh/v3/interp	0.705s
```

Even with 10k iterations, or via the stress tool and many CPUs, I can't seem to get the test to fail in isolation. So maybe it's somehow related to the other tests running.

**Comment by mvdan** [comment]
I think I've tracked down the `signal: hangup` flakes. They essentially boil down to the interpreter getting that error from its stdin/stdout/stderr. The error means "the terminal has been closed".

We were using `creack/pty.Start`, which seems to defer closing the secondary terminal (what is given to `exec.Cmd`):

https://github.com/creack/pty/blob/8ac0cc1360706eb8e5bcd8c0127c63fbdd9d338a/run.go#L50

Unfortunately this seems wrong. If we're simply starting the execution of a command, it might not finish until later. Indeed, if I change the shell source to start with a `sleep 1s; ...`, then the test always fails - by the time it tries to write to stdout, the file has already been closed.

I think this is a bug in upstream. I'm not sure how others haven't run into it. I guess they run commands which are quick enough, and they usually don't notice errors like these. For the time being, I'm just avoiding `pty.Start` and manually using `pty.Open` instead, which seems to entirely get rid of this flake.

I'll close this issue with that fix, as it was by far the most common flake mentioned here. The only other remaining flakes concern redirects; I'm going to be tracking that in https://github.com/mvdan/sh/issues/726.
